### PR TITLE
[RELAY][BACKEND] Enable PlanMemory in the graph runtime.

### PR DESCRIPTION
**Pull request by tqchen**
This PR implements PlanMemory for graph runtime codegen backend of Relay. It also contains a few other improvements

- [x] add show_meta_data option to astext
   - Can be helpful to display text when there is a huge meta-data

The algorithm is basically the same from NNVM. We do need to introduce a storage token and have an initialization phase that propagates and calculate expected reference count of the token, before we run the greedy allocation algorithm.

### CODE REVIEW

**Comment by ajtulloch** [comment]
@tqchen quick question - are there plans to allow dynamic memory allocation in the graph runtime, which would allow variable shapes? I believe that's not currently supported, but was curious if you had plans there.

**Comment by jroesch** [comment]
@ajtulloch yes, I think the plan is to write a new runtime system soon ™️. A few of us are working on a PLDI submission, and expect to ship a bunch of improvements/fixes/features post deadline.

**Comment by jroesch** [comment]
Overall looks good to me, a bit tired from the PLDI push so maybe someone else should do a pass.

**Comment by zhiics** [comment]
@tqchen Quite busy recently. But I will try my best to spend sometime to do a round tonight if it is not too late.

**Comment by tqchen** [comment]
When we start to move into Relay we have this clear separation of compiler and runtime. The migration starts as two-phase process, and we are in the first step that moves the compiler, but keeps the old graph runtime. 

I think we can expect the static graph runtime to exist for a while, but we can also explore the possibility of new backends that breaks different assumptions(e.g. dynamic memory alloca, control flow). Luckily the IR is expressive enough to represent all these workloads. 

There is also a tradeoff here, depending on whether we want to allow JIT, how big the runtime is, etc. So I can imagine it could be possible that we build several of them. @ajtulloch I think it is a good time to hear opinions from everyone on what do we need

**Comment by ajtulloch** [comment]
@tqchen that sounds great. Is there an existing RFC you'd like us to contribute feature suggestions to?

**Comment by tqchen** [comment]
there is not an existing RFC, how about we open a new one?

**Comment by tqchen** [comment]
opened in #2122

**Comment by jroesch** [comment]
@ajtulloch RFC seems like a great idea, would look forward to figuring out what everyone is interested in, and what people are looking to do.

File: src/relay/backend/graph_plan_memory.cc [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+  void VisitExpr_(const GlobalVarNode* op) final {
+    // Do nothing.
+  }
+
+  void VisitExpr_(const OpNode* op) final {
+    // Do nothing.
+  }
+
+  void VisitExpr_(const TupleNode* op) final {
+    // Do nothing.
```
Comment (zhiics) [inline src/relay/backend/graph_plan_memory.cc]: remove the comment?

File: src/relay/backend/graph_plan_memory.cc [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+    token_map_[op->var.operator->()] = token;
+    token_map_[op] = GetToken(op->body);
+  }
+
+ protected:
+  /*! \brief internal token map */
+  std::unordered_map<const ExprNode*, std::vector<StorageToken*> > token_map_;
+
+  /*!
+   * \brief call get token to get the necessary token.
```
Comment (zhiics) [inline src/relay/backend/graph_plan_memory.cc]: Call token

File: src/relay/backend/graph_plan_memory.cc [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    std::vector<StorageToken*> args;
+    // for each input, visit argument token.
+    for (Expr arg : op->args) {
+      for (StorageToken* tok : GetToken(arg)) {
+        args.push_back(tok);
+      }
+    }
+    // create token for the call node.
+    CreateToken(op, true);
+    // check if there is orphaned output that can be released immediately/
```
Comment (zhiics) [inline src/relay/backend/graph_plan_memory.cc]: immediately.

**Comment by zhiics** [comment]
@tqchen I only have some nit comments. Overall LGTM.

File: src/relay/backend/graph_plan_memory.cc [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+#include <tvm/relay/expr_functor.h>
+#include "../../common/arena.h"
+
+namespace tvm {
+namespace relay {
+
+struct StorageToken {
+  /*! \brief Reference counter */
+  int ref_counter{0};
+  /*! \brief numbe of bytes */
```
Comment (yzhliu) [inline src/relay/backend/graph_plan_memory.cc]: typo number

File: src/relay/backend/graph_plan_memory.cc [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+  void VisitExpr_(const FunctionNode* op) final {
+    // do not recursive into sub function.
+  }
+
+  void VisitExpr_(const GlobalVarNode* op) final {
+    // Do nothing.
+  }
+
+  void VisitExpr_(const OpNode* op) final {
+    // Do nothing.
```
Comment (yzhliu) [inline src/relay/backend/graph_plan_memory.cc]: just try to learn, what is the default behavior if such function is not defined?

File: src/relay/backend/graph_plan_memory.cc [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+    CHECK(ttype != nullptr);
+    size_t size = 1;
+    for (IndexExpr dim : ttype->shape) {
+      const int64_t* pval = as_const_int(dim);
+      CHECK(pval != nullptr)
+          << "Cannot allocate memory symbolic tensor shape "
+          << ttype->shape;
+      size *= static_cast<size_t>(pval[0]);
+    }
+    size *= (ttype->dtype.bits() * ttype->dtype.lanes() + 7) / 8;
```
Comment (yzhliu) [inline src/relay/backend/graph_plan_memory.cc]: add comments for magic number 7 & 8?

Comment (ajtulloch) [reply src/relay/backend/graph_plan_memory.cc]: IMO this should be refactored into a round_up/div_round_up function.

Comment (zhiics) [reply src/relay/backend/graph_plan_memory.cc]: +1, it might be necessary to have a alignment function which takes byte, word, or dword, etc.

**Comment by tqchen** [comment]
Thanks, @ajtulloch @yzhliu @zhiics, i have addressed the comments.

**Comment by tqchen** [comment]
Thanks @ajtulloch @yzhliu @zhiics this is merged

**Pull request merged**